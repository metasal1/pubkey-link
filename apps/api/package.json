{
  "name": "api",
  "version": "0.0.1",
  "scripts": {
    "prestart": "pnpm prisma db push",
    "start": "NODE_ENV=production node ./main.js"
  },
  "dependencies": {}
}
