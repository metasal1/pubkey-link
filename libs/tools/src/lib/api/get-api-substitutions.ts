import { names } from '@nx/devkit'
import * as pluralize from 'pluralize'
import { NormalizedApiFeatureSchema } from '../../generators/api-feature/api-feature-schema'

export function getApiSubstitutions(options: NormalizedApiFeatureSchema) {
  const admin = names('admin')
  const model = names(options.modelName)
  const plural = names(pluralize.plural(options.modelName))
  const user = names('user')

  return {
    admin,
    adminCrud: !options.skipAdminCrud,
    adminFileName: admin.fileName,
    modelFileName: model.fileName,
    modelPropertyNamePlural: plural.propertyName,
    app: names(options.app),
    label: names(options.label),
    model,
    npmScope: options.npmScope,
    user,
    // FIXME: Support user crud
    userCrud: false,
  }
}
