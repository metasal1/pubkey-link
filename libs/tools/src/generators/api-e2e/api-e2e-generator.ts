import { Tree } from '@nx/devkit'
import { generateApiE2e } from '../../lib/api-e2e/generate-api-e2e'
import { normalizeApiFeatureSchema } from '../../lib/api/normalize-api-feature-schema'
import { ensureNxProjectExists } from '../../lib/utils/ensure-nx-project-exists'
import { ApiFeatureGeneratorSchema } from '../api-feature/api-feature-schema'

export async function apiE2eGenerator(tree: Tree, rawOptions: ApiFeatureGeneratorSchema) {
  const options = normalizeApiFeatureSchema(tree, rawOptions)
  ensureNxProjectExists(tree, options.app)

  await generateApiE2e(tree, options)
}

export default apiE2eGenerator
