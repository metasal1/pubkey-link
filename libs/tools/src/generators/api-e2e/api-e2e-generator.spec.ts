import { Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { createMockApiApp } from '../../lib/api/create-mock-api-app'
import { ApiFeatureGeneratorSchema } from '../api-feature/api-feature-schema'

import { apiE2eGenerator } from './api-e2e-generator'

describe('api-e2e generator', () => {
  let tree: Tree
  const options: ApiFeatureGeneratorSchema = { app: 'api', name: 'test', label: 'name', skipSdk: true }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it('should run successfully', async () => {
    await createMockApiApp(tree, options.app)
    await apiE2eGenerator(tree, options)

    const basePathE2e = `apps/${options.app}-e2e/src`
    const files = [
      // Generated test
      `${basePathE2e}/api/api-${options.name}-feature.spec.ts`,
    ]

    files.forEach((file) => {
      expect(tree.exists(file)).toBeTruthy()
      expect(tree.read(file).toString()).toMatchSnapshot()
    })
  })
})
