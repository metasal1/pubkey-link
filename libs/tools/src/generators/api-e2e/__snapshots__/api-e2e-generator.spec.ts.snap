// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-e2e generator should run successfully 1`] = `
"import { AdminCreateTestInput, AdminFindManyTestInput, AdminUpdateTestInput, Test } from '@proj/sdk'
import { getAliceCookie, getBobCookie, sdk, uniqueId } from '../support'

describe('api-test-feature', () => {
  describe('api-test-admin-resolver', () => {
    const testName = uniqueId('acme-test')
    let testId: string
    let cookie: string

    beforeAll(async () => {
      cookie = await getAliceCookie()
      const created = await sdk.adminCreateTest({ input: { name: testName } }, { cookie })
      testId = created.data.created.id
    })

    describe('authorized', () => {
      beforeAll(async () => {
        cookie = await getAliceCookie()
      })

      it('should create a test', async () => {
        const input: AdminCreateTestInput = {
          name: uniqueId('test'),
        }

        const res = await sdk.adminCreateTest({ input }, { cookie })

        const item: Test = res.data.created
        expect(item.name).toBe(input.name)
        expect(item.id).toBeDefined()
        expect(item.createdAt).toBeDefined()
        expect(item.updatedAt).toBeDefined()
      })

      it('should update a test', async () => {
        const createInput: AdminCreateTestInput = {
          name: uniqueId('test'),
        }
        const createdRes = await sdk.adminCreateTest({ input: createInput }, { cookie })
        const testId = createdRes.data.created.id
        const input: AdminUpdateTestInput = {
          name: uniqueId('test'),
        }

        const res = await sdk.adminUpdateTest({ testId, input }, { cookie })

        const item: Test = res.data.updated
        expect(item.name).toBe(input.name)
      })

      it('should find a list of tests (find all)', async () => {
        const createInput: AdminCreateTestInput = {
          name: uniqueId('test'),
        }
        const createdRes = await sdk.adminCreateTest({ input: createInput }, { cookie })
        const testId = createdRes.data.created.id

        const input: AdminFindManyTestInput = {}

        const res = await sdk.adminFindManyTest({ input }, { cookie })

        expect(res.data.paging.meta.totalCount).toBeGreaterThan(1)
        expect(res.data.paging.data.length).toBeGreaterThan(1)
        // First item should be the one we created above
        console.log(res.data.paging.data)
        expect(res.data.paging.data[0].id).toBe(testId)
      })

      it('should find a list of tests (find new one)', async () => {
        const createInput: AdminCreateTestInput = {
          name: uniqueId('test'),
        }
        const createdRes = await sdk.adminCreateTest({ input: createInput }, { cookie })
        const testId = createdRes.data.created.id

        const input: AdminFindManyTestInput = {
          search: testId,
        }

        const res = await sdk.adminFindManyTest({ input }, { cookie })

        expect(res.data.paging.meta.totalCount).toBe(1)
        expect(res.data.paging.data.length).toBe(1)
        expect(res.data.paging.data[0].id).toBe(testId)
      })

      it('should find a test by id', async () => {
        const createInput: AdminCreateTestInput = {
          name: uniqueId('test'),
        }
        const createdRes = await sdk.adminCreateTest({ input: createInput }, { cookie })
        const testId = createdRes.data.created.id

        const res = await sdk.adminFindOneTest({ testId }, { cookie })

        expect(res.data.item.id).toBe(testId)
      })

      it('should delete a test', async () => {
        const createInput: AdminCreateTestInput = {
          name: uniqueId('test'),
        }
        const createdRes = await sdk.adminCreateTest({ input: createInput }, { cookie })
        const testId = createdRes.data.created.id

        const res = await sdk.adminDeleteTest({ testId }, { cookie })

        expect(res.data.deleted).toBe(true)

        const findRes = await sdk.adminFindManyTest({ input: { search: testId } }, { cookie })
        expect(findRes.data.paging.meta.totalCount).toBe(0)
        expect(findRes.data.paging.data.length).toBe(0)
      })
    })

    describe('unauthorized', () => {
      let cookie: string
      beforeAll(async () => {
        cookie = await getBobCookie()
      })

      it('should not create a test', async () => {
        expect.assertions(1)
        const input: AdminCreateTestInput = {
          name: uniqueId('test'),
        }

        try {
          await sdk.adminCreateTest({ input }, { cookie })
        } catch (e) {
          expect(e.message).toBe('Unauthorized: User is not Admin')
        }
      })

      it('should not update a test', async () => {
        expect.assertions(1)
        try {
          await sdk.adminUpdateTest({ testId, input: {} }, { cookie })
        } catch (e) {
          expect(e.message).toBe('Unauthorized: User is not Admin')
        }
      })

      it('should not find a list of tests (find all)', async () => {
        expect.assertions(1)
        try {
          await sdk.adminFindManyTest({ input: {} }, { cookie })
        } catch (e) {
          expect(e.message).toBe('Unauthorized: User is not Admin')
        }
      })

      it('should not find a test by id', async () => {
        expect.assertions(1)
        try {
          await sdk.adminFindOneTest({ testId }, { cookie })
        } catch (e) {
          expect(e.message).toBe('Unauthorized: User is not Admin')
        }
      })

      it('should not delete a test', async () => {
        expect.assertions(1)
        try {
          await sdk.adminDeleteTest({ testId }, { cookie })
        } catch (e) {
          expect(e.message).toBe('Unauthorized: User is not Admin')
        }
      })
    })
  })
})
"
`;
